File: weird_captcha_gym/shared_runtime/server/incubator_graders/ribbon_consensus.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "ribbon_consensus"
DEFAULT_PARAMETERS = {
    "row_count": 5,
    "sequence_length": 13,
    "target_gap_count": 2,
    "misalignment_count": 5,
    "max_edits": 8,
    "color_count": 4,
    "mutation_rate": 0.14,
    "gap_open_penalty": 5,
    "gap_extend_penalty": 1,
    "required_gain_ratio": 0.82,
}
CONTROL_FIELDS = frozenset(DEFAULT_PARAMETERS)


def _fail(feedback: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": 0, "feedback": feedback}


def _valid_shift(columns: list[int], break_index: int, delta: int, column_count: int) -> list[int] | None:
    if not 1 <= break_index < len(columns) or delta not in (-1, 1):
        return None
    shifted = list(columns)
    for index in range(break_index, len(shifted)):
        shifted[index] += delta
    if min(shifted) < 0 or max(shifted) >= column_count:
        return None
    if any(left >= right for left, right in zip(shifted, shifted[1:])):
        return None
    return shifted


def score_alignment(
    rows: list[dict[str, Any]],
    columns_by_row: dict[str, list[int]],
    gap_open_penalty: int,
    gap_extend_penalty: int,
) -> dict[str, Any]:
    lookup = {
        row["id"]: {column: index for index, column in enumerate(columns_by_row[row["id"]])}
        for row in rows
    }
    minimum = min(min(values) for values in columns_by_row.values())
    maximum = max(max(values) for values in columns_by_row.values())
    matches = 0
    mismatches = 0
    column_scores: list[int] = []
    for column in range(minimum, maximum + 1):
        tokens: list[int | None] = []
        for row in rows:
            index = lookup[row["id"]].get(column)
            tokens.append(None if index is None else int(row["colors"][index]))
        column_score = 0
        for left_index in range(len(tokens)):
            for right_index in range(left_index + 1, len(tokens)):
                left, right = tokens[left_index], tokens[right_index]
                if left is None or right is None:
                    continue
                if left == right:
                    matches += 1
                    column_score += 1
                else:
                    mismatches += 1
                    column_score -= 1
        column_scores.append(column_score)

    gap_open_runs = 0
    gap_extensions = 0
    for row in rows:
        positions = set(columns_by_row[row["id"]])
        first = min(positions)
        last = max(positions)
        inside_gap = False
        for column in range(first + 1, last):
            if column not in positions:
                if inside_gap:
                    gap_extensions += 1
                else:
                    gap_open_runs += 1
                    inside_gap = True
            else:
                inside_gap = False
    return {
        "score": int(
            matches
            - mismatches
            - gap_open_runs * int(gap_open_penalty)
            - gap_extensions * int(gap_extend_penalty)
        ),
        "matches": int(matches),
        "mismatches": int(mismatches),
        "gap_open_runs": int(gap_open_runs),
        "gap_extensions": int(gap_extensions),
        "column_scores": column_scores,
    }


def _control_contract(
    ground_truth: dict[str, Any], public_state: dict[str, Any]
) -> tuple[dict[str, Any] | None, str | None, str | None]:
    condition = ground_truth.get("control_condition")
    if condition is None:
        if public_state.get("control_condition") is not None:
            raise ValueError("public control condition is unexpected")
        return None, None, "simplified"
    if not isinstance(condition, dict) or public_state.get("control_condition") != condition:
        raise ValueError("control condition differs between public and hidden state")
    difficulty = condition.get("difficulty")
    interaction = str(condition.get("interaction") or "")
    real_time = str(condition.get("real_time") or "")
    parameters = condition.get("difficulty_parameters")
    if (
        isinstance(difficulty, bool)
        or not isinstance(difficulty, int)
        or difficulty not in {1, 2, 3, 4, 5}
        or interaction not in {"simplified", "full"}
        or real_time not in {"live", "paused"}
        or not isinstance(parameters, dict)
        or set(parameters) != CONTROL_FIELDS
    ):
        raise ValueError("control condition is malformed")
    return dict(parameters), interaction, interaction


def _same_json(left: Any, right: Any) -> bool:
    return left == right


def _pointer_path(gesture: dict[str, Any]) -> list[tuple[float, float]] | None:
    path = gesture.get("pointer_path")
    if not isinstance(path, list) or len(path) < 3:
        return None
    points: list[tuple[float, float]] = []
    for point in path:
        if not isinstance(point, dict):
            return None
        try:
            x = float(point["x"])
            y = float(point["y"])
        except (KeyError, TypeError, ValueError, OverflowError):
            return None
        if not math.isfinite(x) or not math.isfinite(y):
            return None
        points.append((x, y))
    return points


def grade(
    payload: dict[str, Any],
    ground_truth: dict[str, Any],
    public_state: dict[str, Any],
) -> dict[str, Any]:
    if not isinstance(payload, dict):
        return _fail("submission is not an object")
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID:
        return _fail("mechanic mismatch")
    if str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return _fail("ground-truth mechanic mismatch")
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id:
        return _fail("stale challenge")
    if str(public_state.get("challenge_id") or "") != challenge_id:
        return _fail("public-state challenge mismatch")
    try:
        parameters, interaction, expected_interaction = _control_contract(ground_truth, public_state)
        parameters = parameters or dict(DEFAULT_PARAMETERS)
        expected_interaction = expected_interaction or "simplified"
        if ground_truth.get("parameters") != parameters or public_state.get("parameters") != parameters:
            raise ValueError("selected parameters are not bound to the generated world")
        if payload.get("task_id") != ground_truth.get("task_id"):
            raise ValueError("task identity mismatch")
        if payload.get("interaction_mode") != expected_interaction:
            raise ValueError("wrong interaction surface")
        if public_state.get("rows") != ground_truth.get("public_rows"):
            raise ValueError("public ribbon tiles differ from the replay contract")
        if public_state.get("initial_columns") != ground_truth.get("initial_columns"):
            raise ValueError("public initial columns differ from the replay contract")
        rows = [dict(row) for row in ground_truth.get("rows") or []]
        if len(rows) != int(parameters["row_count"]):
            raise ValueError("row count does not implement the selected profile")
        stage = dict(ground_truth.get("stage") or {})
        column_count = int(stage["column_count"])
        if int(stage.get("cell_width") or 0) < 24 or int(stage.get("row_height") or 0) < 32:
            raise ValueError("stage geometry is incomplete")
        columns = {
            str(row_id): [int(value) for value in values]
            for row_id, values in (ground_truth.get("initial_columns") or {}).items()
        }
        if set(columns) != {str(row["id"]) for row in rows}:
            raise ValueError("initial column rows are incomplete")
        for row in rows:
            colors = row.get("colors")
            positions = columns[row["id"]]
            if not isinstance(colors, list) or len(colors) != int(parameters["sequence_length"]):
                raise ValueError("row tile count does not implement the selected profile")
            if len(positions) != len(colors) or any(left >= right for left, right in zip(positions, positions[1:])):
                raise ValueError("row tile order is not preserved")
        expected_initial = score_alignment(
            rows,
            columns,
            int(parameters["gap_open_penalty"]),
            int(parameters["gap_extend_penalty"]),
        )
        if public_state.get("score_breakdown") != expected_initial or int(public_state.get("score") or 0) != expected_initial["score"]:
            raise ValueError("initial score is not replayable")
    except (KeyError, TypeError, ValueError, OverflowError) as exc:
        return _fail(f"invalid ribbon contract: {exc}")

    events = payload.get("events")
    max_edits = int(parameters["max_edits"])
    if not isinstance(events, list) or not (1 <= len(events) <= max_edits):
        return _fail("submission has no bounded ribbon edit transcript")
    expected_source = "proxy_shift" if expected_interaction == "simplified" else "tile_run_drag"
    for sequence, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return _fail(f"event {sequence} sequence mismatch")
        if event.get("kind") != "shift" or event.get("input_source") != expected_source:
            return _fail("event uses the wrong input surface")
        if not isinstance(event.get("row_id"), str) or event["row_id"] not in columns:
            return _fail("event names an unknown ribbon")
        try:
            break_index = int(event["break_index"])
            delta = int(event["delta"])
        except (KeyError, TypeError, ValueError, OverflowError):
            return _fail("event shift is malformed")
        shifted = _valid_shift(columns[event["row_id"]], break_index, delta, column_count)
        if shifted is None:
            return _fail("event shifts a tile suffix out of order or off the loom")
        if expected_interaction == "full":
            gesture = event.get("gesture")
            if not isinstance(gesture, dict):
                return _fail("direct tile move has no pointer gesture evidence")
            try:
                sample_count = int(gesture["sample_count"])
                travel_px = float(gesture["travel_px"])
                start_column = int(gesture["start_column"])
                end_column = int(gesture["end_column"])
            except (KeyError, TypeError, ValueError, OverflowError):
                return _fail("pointer gesture evidence is malformed")
            pointer_path = _pointer_path(gesture)
            expected_start_column = columns[event["row_id"]][break_index]
            path_travel = (
                sum(abs(right[0] - left[0]) for left, right in zip(pointer_path, pointer_path[1:]))
                if pointer_path is not None
                else math.nan
            )
            horizontal_delta = (
                pointer_path[-1][0] - pointer_path[0][0]
                if pointer_path is not None
                else math.nan
            )
            if (
                pointer_path is None
                or sample_count != len(pointer_path)
                or sample_count < 3
                or not math.isfinite(travel_px)
                or travel_px < int(stage["cell_width"]) * 0.55
                or not math.isclose(travel_px, path_travel, rel_tol=1e-6, abs_tol=1e-6)
                or abs(horizontal_delta) < int(stage["cell_width"]) * 0.45
                or (horizontal_delta > 0) != (delta > 0)
                or start_column != expected_start_column
                or end_column != expected_start_column + delta
            ):
                return _fail("direct tile move did not contain a physical drag")
        elif event.get("gesture") is not None:
            return _fail("proxy shift includes a direct-pointer gesture")
        columns[event["row_id"]] = shifted
        expected_score = score_alignment(
            rows,
            columns,
            int(parameters["gap_open_penalty"]),
            int(parameters["gap_extend_penalty"]),
        )
        if event.get("score_after") != expected_score or event.get("breakdown_after") != expected_score:
            return _fail(f"event {sequence} score does not replay")

    final_score = score_alignment(
        rows,
        columns,
        int(parameters["gap_open_penalty"]),
        int(parameters["gap_extend_penalty"]),
    )
    try:
        score_floor = int(ground_truth["score_floor"])
        submitted_score = int(payload["final_score"])
        move_count = int(payload["move_count"])
    except (KeyError, TypeError, ValueError, OverflowError):
        return _fail("final ribbon certificate is malformed")
    if (
        payload.get("final_columns") != columns
        or submitted_score != final_score["score"]
        or payload.get("score_breakdown") != final_score
        or move_count != len(events)
        or payload.get("completed") is not True
        or final_score["score"] < score_floor
    ):
        return _fail(
            f"alignment not certified: score {final_score['score']} / required {score_floor}; "
            f"moves {len(events)}/{max_edits}"
        )
    return {
        "graded": True,
        "passed": True,
        "score": 100,
        "feedback": f"alignment certified at score {final_score['score']} after {len(events)} suffix shifts",
        "final_score": final_score,
    }


def cheat(ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    """Development-only witness for the ordinary browser solver; never shown in task UI."""
    return {
        "mechanic_id": MECHANIC_ID,
        "challenge_id": ground_truth.get("challenge_id"),
        "interaction": (ground_truth.get("control_condition") or {}).get("interaction", "simplified"),
        "repair_actions": ground_truth.get("repair_actions") or [],
        "target_columns": ground_truth.get("target_columns") or {},
        "target_score": ground_truth.get("target_score") or {},
        "public_challenge_id": public_state.get("challenge_id"),
    }

File: weird_captcha_gym/shared_runtime/app/mechanics/ribbon_consensus.js
(() => {
  "use strict";

  const model = {
    state: null,
    helpers: null,
    interaction: "simplified",
    columns: {},
    events: [],
    score: null,
    drag: null,
    busy: false,
    terminal: false,
  };

  const clean = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
  const clamp = (value, low, high) => Math.max(low, Math.min(high, value));

  function rowsForScore() {
    return (model.state.rows || []).map((row) => ({
      id: row.id,
      colors: (row.tiles || []).map((tile) => Number(tile.color)),
    }));
  }

  function scoreAlignment() {
    const rows = rowsForScore();
    const lookup = Object.fromEntries(rows.map((row) => [
      row.id,
      Object.fromEntries((model.columns[row.id] || []).map((column, index) => [column, index])),
    ]));
    const allColumns = Object.values(model.columns).flat();
    const minimum = Math.min(...allColumns);
    const maximum = Math.max(...allColumns);
    let matches = 0;
    let mismatches = 0;
    const columnScores = [];
    for (let column = minimum; column <= maximum; column += 1) {
      const tokens = rows.map((row) => {
        const index = lookup[row.id][column];
        return index == null ? null : row.colors[index];
      });
      let columnScore = 0;
      for (let leftIndex = 0; leftIndex < tokens.length; leftIndex += 1) {
        for (let rightIndex = leftIndex + 1; rightIndex < tokens.length; rightIndex += 1) {
          const left = tokens[leftIndex];
          const right = tokens[rightIndex];
          if (left == null || right == null) continue;
          if (left === right) {
            matches += 1;
            columnScore += 1;
          } else {
            mismatches += 1;
            columnScore -= 1;
          }
        }
      }
      columnScores.push(columnScore);
    }
    let gapOpenRuns = 0;
    let gapExtensions = 0;
    rows.forEach((row) => {
      const positions = new Set(model.columns[row.id]);
      const first = Math.min(...positions);
      const last = Math.max(...positions);
      let insideGap = false;
      for (let column = first + 1; column < last; column += 1) {
        if (!positions.has(column)) {
          if (insideGap) gapExtensions += 1;
          else {
            gapOpenRuns += 1;
            insideGap = true;
          }
        } else {
          insideGap = false;
        }
      }
    });
    const parameters = model.state.parameters;
    return {
      score: matches - mismatches - gapOpenRuns * Number(parameters.gap_open_penalty) - gapExtensions * Number(parameters.gap_extend_penalty),
      matches,
      mismatches,
      gap_open_runs: gapOpenRuns,
      gap_extensions: gapExtensions,
      column_scores: columnScores,
    };
  }

  function validShift(columns, breakIndex, delta) {
    if (!Number.isInteger(breakIndex) || breakIndex < 1 || breakIndex >= columns.length || ![-1, 1].includes(delta)) return null;
    const shifted = [...columns];
    for (let index = breakIndex; index < shifted.length; index += 1) shifted[index] += delta;
    const columnCount = Number(model.state.column_count);
    if (Math.min(...shifted) < 0 || Math.max(...shifted) >= columnCount) return null;
    for (let index = 1; index < shifted.length; index += 1) if (shifted[index - 1] >= shifted[index]) return null;
    return shifted;
  }

  function setReadout(message, status = "idle") {
    model.helpers.setReadout(message, status);
  }

  function updateSelectOptions() {
    const rowSelect = document.querySelector("#ribbon-row-select");
    const breakSelect = document.querySelector("#ribbon-break-select");
    if (!rowSelect || !breakSelect) return;
    const oldRow = rowSelect.value || model.state.rows[0]?.id;
    rowSelect.innerHTML = (model.state.rows || []).map((row) => `<option value="${clean(row.id)}">${clean(row.label)}</option>`).join("");
    rowSelect.value = model.state.rows.some((row) => row.id === oldRow) ? oldRow : model.state.rows[0]?.id;
    const row = model.state.rows.find((item) => item.id === rowSelect.value) || model.state.rows[0];
    const oldBreak = Number(breakSelect.value || 1);
    breakSelect.innerHTML = (row?.tiles || []).slice(1).map((tile) => `<option value="${tile.index}">after tile ${tile.index}</option>`).join("");
    const maximum = Math.max(1, (row?.tiles || []).length - 1);
    breakSelect.value = String(clamp(oldBreak, 1, maximum));
  }

  function updatePanel() {
    model.score = scoreAlignment();
    const scoreNode = document.querySelector(".ribbon-score-value");
    const floorNode = document.querySelector(".ribbon-score-floor");
    const movesNode = document.querySelector(".ribbon-moves-value");
    const matchNode = document.querySelector(".ribbon-match-value");
    const gapNode = document.querySelector(".ribbon-gap-value");
    if (scoreNode) scoreNode.textContent = String(model.score.score);
    if (floorNode) floorNode.textContent = `required ${model.state.score_floor}`;
    if (movesNode) movesNode.textContent = `${model.events.length} / ${model.state.max_edits} edits`;
    if (matchNode) matchNode.textContent = `${model.score.matches} matches · ${model.score.mismatches} mismatches`;
    if (gapNode) gapNode.textContent = `${model.score.gap_open_runs} opens · ${model.score.gap_extensions} extensions`;
    const progress = document.querySelector(".ribbon-score-progress i");
    if (progress) {
      const start = Number(model.state.score || 0);
      const required = Number(model.state.score_floor || 0);
      const fraction = required === start ? (model.score.score >= required ? 1 : 0) : (model.score.score - start) / (required - start);
      progress.style.width = `${clamp(fraction * 100, 0, 100)}%`;
    }
    document.querySelectorAll(".ribbon-column-score").forEach((node, index) => {
      const value = model.score.column_scores[index] || 0;
      node.textContent = value > 0 ? `+${value}` : String(value);
      node.dataset.tone = value > 0 ? "good" : value < 0 ? "bad" : "flat";
    });
    const ready = model.score.score >= Number(model.state.score_floor);
    const badge = document.querySelector(".ribbon-cert-badge");
    if (badge) {
      badge.textContent = ready ? "THRESHOLD REACHED · CERTIFY" : "ALIGNMENT IN PROGRESS";
      badge.dataset.ready = ready ? "true" : "false";
    }
    const exhausted = model.events.length >= Number(model.state.max_edits);
    setReadout(ready ? "THRESHOLD REACHED · CERTIFY THE LOOM" : exhausted ? "EDIT LIMIT REACHED · CERTIFY FOR A FRESH LOOM" : "SHIFT A TILE RUN TO IMPROVE COLUMN AGREEMENT", ready ? "ready" : exhausted ? "error" : "idle");
  }

  function tileMarkup(row, tile, rowIndex) {
    const color = model.state.colors[Number(tile.color)] || {};
    const column = model.columns[row.id][Number(tile.index)];
    const cell = Number(model.state.stage.cell_width);
    const labelWidth = 108;
    return `<button type="button" class="loom-tile" data-row-id="${clean(row.id)}" data-index="${tile.index}" aria-label="${clean(row.label)} tile ${Number(tile.index) + 1}, ${clean(tile.color_name)}" style="left:${labelWidth + column * cell + 3}px;top:6px;--tile-color:${clean(color.hex)};--tile-ink:${clean(color.ink)}"><span>${Number(tile.index) + 1}</span></button>`;
  }

  function renderBoard() {
    const board = document.querySelector(".loom-board");
    if (!board) return;
    board.innerHTML = (model.state.rows || []).map((row, rowIndex) => `
      <div class="loom-row" style="top:${rowIndex * Number(model.state.stage.row_height)}px;height:${Number(model.state.stage.row_height)}px">
        <div class="loom-row-label"><span>${clean(row.label)}</span><small>${row.tiles.length} TILES</small></div>
        <div class="loom-track"></div>
        ${(row.tiles || []).map((tile) => tileMarkup(row, tile, rowIndex)).join("")}
      </div>`).join("") + `<div class="loom-column-audit">${Array.from({length: Number(model.state.column_count)}, (_, index) => `<span class="ribbon-column-score" style="left:${108 + index * Number(model.state.stage.cell_width) + 2}px">0</span>`).join("")}</div>`;
    if (model.interaction !== "full") return;
    board.querySelectorAll(".loom-tile").forEach((tile) => {
      tile.addEventListener("pointerdown", (event) => {
        if (model.busy || model.terminal) return;
        const rowId = tile.dataset.rowId;
        const breakIndex = Number(tile.dataset.index);
        if (breakIndex < 1) {
          setReadout("GRAB A TILE AFTER THE FIRST · SUFFIX SHIFTS PRESERVE ORDER", "error");
          return;
        }
        const startColumn = Number(model.columns[rowId]?.[breakIndex]);
        if (!Number.isInteger(startColumn)) {
          setReadout("THAT TILE IS NOT AVAILABLE IN THE CURRENT LOOM", "error");
          return;
        }
        model.drag = {
          rowId,
          breakIndex,
          startColumn,
          startX: event.clientX,
          lastX: event.clientX,
          travelPx: 0,
          pointerPath: [{x: event.clientX, y: event.clientY}],
          pointerId: event.pointerId,
        };
        tile.classList.add("is-dragging");
        tile.setPointerCapture?.(event.pointerId);
        event.preventDefault();
      });
    });
    board.onpointermove = (event) => {
      if (!model.drag || event.pointerId !== model.drag.pointerId) return;
      model.drag.travelPx += Math.abs(event.clientX - model.drag.lastX);
      model.drag.lastX = event.clientX;
      model.drag.pointerPath.push({x: event.clientX, y: event.clientY});
    };
    board.onpointerup = (event) => finishDrag(event);
    board.onpointercancel = (event) => finishDrag(event, true);
  }

  function finishDrag(event, cancelled = false) {
    if (!model.drag || event.pointerId !== model.drag.pointerId) return;
    const drag = model.drag;
    model.drag = null;
    document.querySelectorAll(".loom-tile.is-dragging").forEach((node) => node.classList.remove("is-dragging"));
    if (cancelled) return;
    const cell = Number(model.state.stage.cell_width);
    const finalX = Number(event.clientX);
    const finalY = Number(event.clientY);
    drag.travelPx += Math.abs(finalX - drag.lastX);
    drag.pointerPath.push({x: finalX, y: finalY});
    const distance = finalX - drag.startX;
    const delta = Math.abs(distance) >= cell * 0.45 ? (distance > 0 ? 1 : -1) : 0;
    if (!delta || drag.travelPx < cell * 0.55 || drag.pointerPath.length < 3) {
      setReadout("DRAG ONE CELL LEFT OR RIGHT TO SHIFT THE TILE RUN", "error");
      return;
    }
    applyShift(drag.rowId, drag.breakIndex, delta, {
      sample_count: drag.pointerPath.length,
      travel_px: drag.travelPx,
      start_column: drag.startColumn,
      end_column: drag.startColumn + delta,
      pointer_path: drag.pointerPath,
    });
  }

  function applyShift(rowId, breakIndex, delta, gesture = null) {
    if (model.busy || model.terminal) return;
    if (model.events.length >= Number(model.state.max_edits)) {
      setReadout("EDIT LIMIT REACHED · CERTIFY TO FINISH OR RECEIVE A FRESH LOOM", "error");
      return;
    }
    document.querySelector(".ribbon-verdict-fail")?.remove();
    document.querySelector(".ribbon-consensus")?.classList.remove("is-fresh-fail");
    const shifted = validShift(model.columns[rowId], breakIndex, delta);
    if (!shifted) {
      setReadout("THAT RUN WOULD COLLIDE WITH THE LOOM EDGE OR BREAK ORDER", "error");
      return;
    }
    model.columns[rowId] = shifted;
    const after = scoreAlignment();
    model.events.push({
      sequence: model.events.length + 1,
      kind: "shift",
      row_id: rowId,
      break_index: breakIndex,
      delta,
      input_source: model.interaction === "full" ? "tile_run_drag" : "proxy_shift",
      ...(gesture ? {gesture} : {}),
      score_after: after,
      breakdown_after: after,
    });
    renderBoard();
    updateSelectOptions();
    updatePanel();
  }

  async function submit() {
    if (model.busy || model.terminal) return;
    model.busy = true;
    document.querySelectorAll("button,select").forEach((node) => { node.disabled = true; });
    const finalScore = scoreAlignment();
    setReadout("CERTIFYING THE RIBBON CONSENSUS…", "pending");
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      interaction_mode: model.interaction,
      events: model.events,
      final_columns: model.columns,
      final_score: finalScore.score,
      score_breakdown: finalScore,
      move_count: model.events.length,
      completed: finalScore.score >= Number(model.state.score_floor),
    };
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify(payload),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".ribbon-consensus")?.classList.add("is-pass");
        document.querySelector(".ribbon-consensus")?.insertAdjacentHTML("beforeend", '<div class="ribbon-verdict ribbon-verdict-pass"><small>LOOM CERTIFICATE ACCEPTED</small><strong>PASS</strong></div>');
        setReadout("PASS · RIBBON CONSENSUS CERTIFIED", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const shell = document.querySelector(".ribbon-consensus");
        shell?.classList.add("is-fresh-fail");
        shell?.insertAdjacentHTML("beforeend", '<div class="ribbon-verdict ribbon-verdict-fail"><small>ALIGNMENT REJECTED · FRESH LOOM ISSUED</small><strong>FAIL</strong></div>');
        model.helpers.setReadout("FAIL · FRESH LOOM ISSUED", "error");
      } else {
        model.busy = false;
        document.querySelectorAll("button,select").forEach((node) => { node.disabled = false; });
        setReadout("FAIL · CERTIFICATION SERVICE OFFLINE", "error");
      }
    } catch (_error) {
      model.busy = false;
      document.querySelectorAll("button,select").forEach((node) => { node.disabled = false; });
      setReadout("FAIL · CERTIFICATION SERVICE OFFLINE", "error");
    }
  }

  function bindControls() {
    const rowSelect = document.querySelector("#ribbon-row-select");
    rowSelect?.addEventListener("change", updateSelectOptions);
    document.querySelector("[data-shift='1']")?.addEventListener("click", () => {
      applyShift(rowSelect.value, Number(document.querySelector("#ribbon-break-select").value), 1);
    });
    document.querySelector("[data-shift='-1']")?.addEventListener("click", () => {
      applyShift(rowSelect.value, Number(document.querySelector("#ribbon-break-select").value), -1);
    });
    document.querySelector(".ribbon-submit")?.addEventListener("click", submit);
  }

  async function render(state, helpers) {
    document.body.dataset.mechanic = "ribbon-consensus";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    model.state = state;
    model.helpers = helpers;
    model.interaction = String((state.control_condition || {}).interaction || "simplified");
    model.columns = Object.fromEntries(Object.entries(state.initial_columns || {}).map(([key, value]) => [key, [...value]]));
    model.events = [];
    model.score = null;
    model.drag = null;
    model.busy = false;
    model.terminal = false;
    const colorLegend = (state.colors || []).map((color) => `<span class="ribbon-color-key"><i style="background:${clean(color.hex)}"></i>${clean(color.name)}</span>`).join("");
    const interactionCopy = model.interaction === "full"
      ? "Drag any tile one loom cell. Its suffix follows, preserving ribbon order."
      : "Choose a ribbon and boundary, then use the visible shift controls. The same suffix move is replayed.";
    helpers.app.innerHTML = `<section class="ribbon-consensus" data-interaction="${clean(model.interaction)}" data-challenge-id="${clean(state.challenge_id)}">
      <header class="ribbon-head">
        <div><span class="ribbon-kicker">OPEN-PHYLO / CONSENSUS LOOM</span><h1>${clean(state.prompt)}</h1><p>Every colored tile keeps its ribbon order. Only a suffix may move across a column gap.</p></div>
        <div class="ribbon-target-card"><small>CERTIFICATION FLOOR</small><strong>${clean(state.score_floor)}</strong><span>live sum-of-pairs score</span></div>
      </header>
      <section class="ribbon-legend"><span class="ribbon-legend-title">VISIBLE TILE PALETTE</span>${colorLegend}<span class="ribbon-id">LOOM ${clean(state.challenge_id)}</span></section>
      <main class="ribbon-main">
        <section class="loom-panel">
          <div class="loom-panel-head"><div><span>WEAVE FIELD / ${clean(state.rows.length)} RIBBONS</span><h2>Align the vertical consensus bands</h2></div><b class="ribbon-cert-badge" data-ready="false">ALIGNMENT IN PROGRESS</b></div>
          <div class="loom-scroll"><div class="loom-board" style="--loom-columns:${state.column_count};--loom-cell:${state.stage.cell_width}px;--loom-row-height:${state.stage.row_height}px;width:${108 + state.column_count * state.stage.cell_width}px;height:${state.rows.length * state.stage.row_height + 20}px"></div></div>
          <div class="loom-column-key"><span>VERTICAL COLUMN AUDIT</span><small>positive columns reward agreement · gap penalties are included in score</small></div>
        </section>
        <section class="ribbon-console-row">
          <aside class="ribbon-console">
            <span class="console-kicker">${model.interaction === "full" ? "DIRECT LOOM CONTROL" : "BOUNDARY CONTROL"}</span><h2>${model.interaction === "full" ? "Drag a tile run" : "Shift a tile run"}</h2><p>${interactionCopy}</p>
            ${model.interaction === "simplified" ? `<label>RIBBON<select id="ribbon-row-select"></select></label><label>BOUNDARY<select id="ribbon-break-select"></select></label><div class="ribbon-shift-buttons"><button type="button" data-shift="-1">← CLOSE GAP</button><button type="button" data-shift="1">OPEN GAP →</button></div>` : `<div class="direct-hint"><i>1</i><span>Press on any tile after the first.</span><i>2</i><span>Drag one cell left or right.</span><i>3</i><span>Release to replay the suffix shift.</span></div>`}
            <div class="ribbon-rule"><b>ORDER LOCK</b><span>Tiles never reorder; a move shifts this tile and everything after it in the row.</span></div>
          </aside>
          <aside class="ribbon-score-card"><span class="console-kicker">LIVE COLUMN FEEDBACK</span><div class="ribbon-score-line"><strong class="ribbon-score-value">0</strong><small class="ribbon-score-floor"></small></div><div class="ribbon-score-progress"><i></i></div><div class="ribbon-score-stats"><span class="ribbon-match-value"></span><span class="ribbon-gap-value"></span><span class="ribbon-moves-value"></span></div><div class="ribbon-score-note">The floor is reachable from this loom, but the target alignment is not shown.</div></aside>
        </section>
      </main>
      <footer class="ribbon-foot"><div><span>CONSENSUS STATUS</span><div class="readout" data-status="idle">SHIFT A TILE RUN TO IMPROVE COLUMN AGREEMENT</div></div><button type="button" class="ribbon-submit">${clean(state.submit_label || "CERTIFY ALIGNMENT")}</button></footer>
      ${helpers.cheatPanelTemplate()}
    </section>`;
    renderBoard();
    updateSelectOptions();
    bindControls();
    updatePanel();
    helpers.installCheatPanel();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.ribbon_consensus = {render};
})();
